C SDK - Remove static buffer from serializer type system

obsoleted opened the pull request:
The tempbuffer used in the type system is quite large for some of the target devices we are trying to port the C sdk to. This change updates the client to dynamically allocate temporary buffers as needed. Each new usage now makes an estimate of the size needed and allocates that. Unfortunately exact size determination didn't seem straightforward in a portable way.

- Removed the static (10k!) buffer from agenttypesystem.c
- Moved to a dynamic allocation model for character buffer as needed
- Estimated relatively conservative upper limit string sizes for printing
  longs and floats.
- All tests in azure_iot_sdks.sln currently passing

azurecla commented:
Hi __@obsoleted__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

dcristoloveanu commented on c/serializer/src/agenttypesystem.c:
```diff
@@ -42,7 +42,25 @@
+// An example of this would be printf("%.*f", MAX_FLOATING_POINT_STRING_LENGTH, 1.3);
+// But currently no explicit requests for this exist in the file nor are
+// any expected to reasonably occur when being used (numbers that hit
+// this limit would be experiencing signifcant precision loss in storage anyway.
```
Typo:significant

dcristoloveanu commented on c/serializer/src/agenttypesystem.c:
```diff
@@ -1417,10 +1458,40 @@ AGENT_DATA_TYPES_RESULT AgentDataTypes_ToString(STRING_HANDLE destination, const
+                        // Clean up temp buffer if allocated
+                        if (tempBuffer != NULL) {
+                            free(tempBuffer);
+                            tempBuffer = NULL;
```
I think you can skip the "else if"in this case and have  {} for the else statement. And then a one liner free on the else branch of the allocation in line 1433 would do it. Then we don't have a second check here whether tempBuffer has been allocated.

dcristoloveanu commented on c/serializer/src/agenttypesystem.c:
```diff
@@ -1394,7 +1409,33 @@ AGENT_DATA_TYPES_RESULT AgentDataTypes_ToString(STRING_HANDLE destination, const
+                            1 + //\"
+                            1;  // " (terminating NULL);
+
+                        char* tempBuffer = malloc(tempBufferSize);
```
A cast to (char*) would help here if compiling the code as C++ (we had some request for that at some point).

dcristoloveanu commented on c/serializer/src/agenttypesystem.c:
```diff
@@ -1469,10 +1571,39 @@ AGENT_DATA_TYPES_RESULT AgentDataTypes_ToString(STRING_HANDLE destination, const
                             result = AGENT_DATA_TYPES_OK;
                         }
+
+                        if (tempBuffer != NULL)
```
Same comment about doing the free on the else branch of the check in ln 1548.

dcristoloveanu commented:
Fix looks good, only some minor comments and we'll pull it in.
I assume tests are still passing, since I don't see any API changes on AgentTypeSystem.

obsoleted commented:
Thanks! 

I've added a commit that hopefully addresses your feedback. 

All tests from the build command are passing for me:
```
...
100% tests passed, 0 tests failed out of 67

Total Test time (real) = 145.14 sec
```

obsoleted commented:
I also ran all the tests in the azure_iot_sdks solution. All 3980 discovered tests passed.

dcristoloveanu commented:
This has been merged in.
